lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+    { shareStateAccountKey :: k 'AccountK XPub
+        -- ^ Reward account public key associated with this wallet
+    , shareStateDerivationPrefix :: !DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+    , shareStateGap :: !AddressPoolGap
+        -- ^ Number of first keys that are used to produce candidate addresses
+    , shareStatePaymentTemplate :: !ScriptTemplate
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for payment credential
+    , shareStateDelegationTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for staking credential. If not specified then the same template as for
+        -- payment is used
+    , shareStateOurAddresses :: !(Map Address (Index 'Soft 'ScriptK, AddressState))
+        -- ^ Our addresses meaning the addresses containing our verification key hashes
+        -- (if discovered) and represented here by corresponding indices of verification
+        -- keys present in the given address.
+    }
+    deriving stock (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    ) => Show (SharedState n k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    ) => Eq (SharedState n k)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    ) => NFData (SharedState n k)
+
+-- | Purpose for shared wallets is a constant set to 45' (or 0x8000002D) following the original
+-- CIP-XXX Multi-signature Wallets.
+--
+-- It indicates that the subtree of this node is used according to this
+-- specification.
+--
+-- Hardened derivation is used at this level.
+purposeCIPXXX :: Index 'Hardened 'PurposeK
+purposeCIPXXX = toEnum 0x8000002D
+
+-- | Create a SharedState from the all needed ingredients.
+-- There is no validation and it is unsafe way.
+unsafeSharedState
+    :: forall (n :: NetworkDiscriminant) k. k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Map Address (Index 'Soft 'ScriptK, AddressState)
+    -> SharedState n k
+unsafeSharedState accXPub accIx g pTemplate dTemplate ourAdresesses =
+    SharedState
+    { shareStateAccountKey = accXPub
+    , shareStateDerivationPrefix = DerivationPrefix ( purposeCIPXXX, coinTypeAda, accIx )
+    , shareStateGap = g
+    , shareStatePaymentTemplate = pTemplate
+    , shareStateDelegationTemplate = dTemplate
+    , shareStateOurAddresses = ourAdresesses
+    }
+
+-- | Create a new SharedState.
+newSharedState
+    :: forall (n :: NetworkDiscriminant) k. k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+newSharedState accXPub accIx g pTemplate dTemplate =
+    extendSharedState (toEnum 0) $
+    unsafeSharedState accXPub accIx g pTemplate dTemplate Map.empty
+
+replaceCosignersWithVerKeys
+    :: ScriptTemplate
+    -> Int
```

Comment:
We can do better than `Int` here :) --> `Index 'Soft 'ScriptK` sounds like a good candidate.

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/9522d0c0a5b0fd39ffab1b03235f364c78a972f3